TLM/TLM/Manager/Impl/RoutingManager.cs

```diff
@@ -441,89 +441,131 @@ public class RoutingManager : AbstractGeometryObservingManager, IRoutingManager
+                LaneTransitionData[] nextForcedTransitionDatas = null;
+                byte numNextForcedTransitionDatas = 0;
+                int[] nextCompatibleTransitionDataIndices = null;
+                byte numNextCompatibleTransitionDataIndices = 0;
+                int[] compatibleLaneIndexToLaneConnectionIndex = null;
+
+                if (isNextSegmentValid) {
+                    nextRelaxedTransitionDatas = new LaneTransitionData[MAX_NUM_TRANSITIONS];
+                    nextCompatibleTransitionDatas = new LaneTransitionData[MAX_NUM_TRANSITIONS];
+                    nextLaneConnectionTransitionDatas = new LaneTransitionData[MAX_NUM_TRANSITIONS];
+                    nextForcedTransitionDatas = new LaneTransitionData[MAX_NUM_TRANSITIONS];
+                    nextCompatibleOuterSimilarIndices = new int[MAX_NUM_TRANSITIONS];
+                    nextCompatibleTransitionDataIndices = new int[MAX_NUM_TRANSITIONS];
+                    compatibleLaneIndexToLaneConnectionIndex = new int[MAX_NUM_TRANSITIONS];
+                }
+
+
+                uint nextLaneId = nextFirstLaneId;
+                byte nextLaneIndex = 0;
+                //ushort compatibleLaneIndicesMask = 0;
+                int[] laneGroups = new int[nextSegmentInfo.m_lanes.Length];
+                int prevLaneGroup = 0;
+                List<NetInfo.Lane> prevSortedLanes = prevSegmentInfo.m_lanes.Where(l => (l.m_laneType & LaneConnectionManager.LANE_TYPES) != NetInfo.LaneType.None
+                    && (l.m_vehicleType & LaneConnectionManager.VEHICLE_TYPES) != VehicleInfo.VehicleType.None).OrderBy(l => l.m_position).ToList();
+                List<NetInfo.Lane> sortedLanes = nextSegmentInfo.m_lanes.Where(l => (l.m_laneType & LaneConnectionManager.LANE_TYPES) != NetInfo.LaneType.None
+                    && (l.m_vehicleType & LaneConnectionManager.VEHICLE_TYPES) != VehicleInfo.VehicleType.None).OrderBy(l => l.m_position).ToList();
+
+                if (prevSegmentInfo == nextSegmentInfo) {
+                    if (!nextIsJunction) {
+
+                        int prevLaneSortedIndex = prevSortedLanes.IndexOf(prevLaneInfo);
+                        for (byte prevLaneInx = 1; prevLaneInx <= prevLaneSortedIndex && prevLaneId != 0; prevLaneInx++) {
+                            NetInfo.Lane laneInfo = prevSortedLanes[prevLaneInx];
+                            if ((laneInfo.m_vehicleType & VehicleInfo.VehicleType.Car) == VehicleInfo.VehicleType.None) {
+                                prevLaneGroup++;
+                            }
+                            if (laneInfo.m_finalDirection != prevSortedLanes[prevLaneInx - 1].m_finalDirection) {
+                                prevLaneGroup++;
+                            } else {
+                                var gap = laneInfo.m_position - prevSortedLanes[prevLaneInx - 1].m_position - (0.5f * prevSortedLanes[prevLaneInx - 1].m_width) - (0.5f * laneInfo.m_width);
```

Comment:
Is a gap really a good indicator for lane accessibility? Why not checking whether adjacent lanes have a compatible vehicle type?

Reply:
And if that's not sufficient: I believe roads with medians have some kind of `m_canCrossLanes` flag. But I don't know if asset creators have used it.

Reply:
IMO the `m_canCrossLanes` flag is not related to what this PR is trying to achieve. I discussed that flag with a few road asset authors - they tried using it but it didn't seem to suffice. From what they were saying the flag determines whether vehicle can cross oncoming traffic to get to building on other side of road, and the flag seems mainly aimed at the small or medium roads. It doesn't seem to have any effect on vehicles ability, or lack of, to cross medians. See #89 for examples of some of the roads in question.

Reply:
Ok, then forget this flag. xD It would not work for multiple medians either.

Reply:
Yes, it will need some trial and error to get the correct gap size. That being said, I would imagine the width of a vanilla median is a good reference point - @andreharv could you confirm?

Also, is there a way to find out what's "in" a gap? For example, if it's pavement of some kind, we know it's definitely a line not to cross. But if it's empty space (like dashed line strip down the middle of a road) that would be useful for the emergency vehicle stuff that's being developed (eg. emergency vehicles could use it, or cars could use it as temporary parking to allow emergency vehicles to get through).

BTW, there are some unusual roads that might make good test cases listed in #34 and #30.

Reply:
Thinking more on this, the `m_canCrossLanes` flag would also need to respect the medians, so vehicles can only cross oncoming traffic if `m_canCrossLanes == true` and there are no medians in the way.

> Aside: Would it be useful to have a page detailing the various road flags and what they do, from perspective of TMPE? CSL Modding Info website has the information from asset creator perspective, but I've seen little to nothing from the code/modding perspective as to what effects these flags actually have in pathfinder and vehicle AIs.

Reply:
> Also, is there a way to find out what's "in" a gap?

Like I said: We could check whether adjacent lanes have different vehicle types assigned and use that as a primary estimator. But when an impassable track is not represented by a `NetLane`, we need to have a fallback mechnasim (like the gap check).

Reply:
Could you clarify what you mean by 'different vehicle types assigned'? Is that vehicle restrictions added by TMPE, or vehicle classes defined by base game, or something else?

I would agree in the case of, for example, a pedestrian lane (which are sometimes present on medians) but if it was just different type of road vehicle it might cause false positives. Also, we have to view this from the perspective of other types of vehicles - for example, pedestrians perceive road lanes and train tracks to be impassible medians.

Also for gap checks, is there any way to cache the info per prefab so we can avoid having to do loads of checking at runtime? If I'm reading the code correctly (which may not be the case as I'm still total noob at C#) it looks like it's trying to determine where medians are while pathfinding rather than doing some one-off up-front processing when a level is loaded. If we detect a gap, we could add a fake 'median lane' or something?

Reply:
> Could you clarify what you mean by 'different vehicle types assigned'? Is that vehicle restrictions added by TMPE, or vehicle classes defined by base game, or something else?
That's coming from the base game. Each lane info has a lane type (`m_laneType`) and (`m_vehicleType`) assigned. A regular car lane is represented by Vehicle/Car (lane type/vehicle type), a bus lane by TransportVehicle/Car, a tram track by Vehicle/Tram and a sidewalk by Pedestrian/None.

> but if it was just different type of road vehicle it might cause false positives
When could you get false positives? If asset creators model those lanes, checking vehicle types of adjacent lanes should work. If not, we have to check for gaps.

> Also for gap checks, is there any way to cache the info per prefab so we can avoid having to do loads of checking at runtime?
👍